File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        sums = ListNode()
        first = sums
        while l1 or l2:
            if l1 and l2:
                if l1.val + l2.val + sums.val >= 10:
                    sums.next = ListNode(1)
                sums.val = (l1.val + l2.val + sums.val) % 10  
                l1 = l1.next
                l2 = l2.next
            elif l1:
                if l1.val + sums.val >= 10:
                    sums.next = ListNode(1)
                sums.val = (l1.val + sums.val) % 10  
                l1 = l1.next
            else:
                if l2.val + sums.val >= 10:
                    sums.next = ListNode(1)
                sums.val = (l2.val + sums.val) % 10  
                l2 = l2.next
            if l1 or l2:
                if not sums.next:
                    sums.next = ListNode(0)
                sums = sums.next
        return first

        

File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {

    struct ListNode* sums = malloc(sizeof(struct ListNode));
    struct ListNode* original = sums;
    int carry = 0;
    sums->val = 0;
    sums->next = NULL;
    while (l1 || l2 || carry) {
        int sum = 0;
        sum += carry;
        if (l1) {
            sum += l1->val;
            l1 = l1->next;
        }

        if (l2) {
            sum += l2->val;
            l2 = l2->next;
        }

        
        carry = sum / 10;
        sums->val = sum % 10;
        if (l1 || l2 || carry) {
            sums->next = malloc(sizeof(struct ListNode));
            sums = sums->next;
            sums->val = 0;
            sums->next = NULL;
        }
    }
    return original;
}